File: src/statement.ts
import type Token from './token';
import type { Expression, VariableExpression } from './expression';
import type { LiteralType } from './type';
export interface StatementVisitor {
  visitBlockStatement: (statement: BlockStatement) => LiteralType;
  visitClassStatement: (statement: ClassStatement) => LiteralType;
  visitExpressionStatement: (statement: ExpressionStatement) => LiteralType;
  visitFunctionStatement: (statement: FunctionStatement) => LiteralType;
  visitIfStatement: (statement: IfStatement) => LiteralType;
  visitReturnStatement: (statement: ReturnStatement) => LiteralType;
  visitVariableStatement: (statement: VariableStatement) => LiteralType;
  visitWhileStatement: (statement: WhileStatement) => LiteralType;
}
export interface Statement {
  accept(visitor: StatementVisitor): LiteralType;
  toString(): string;
}
export class BlockStatement implements Statement {
  readonly statements: Statement[];
  constructor(statements: Statement[]) {
    this.statements = statements;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitBlockStatement(this);
  }
  toString() {
    return `{${this.statements.map((item) => item.toString()).join('')}}`;
  }
}
export class ClassStatement implements Statement {
  readonly name: Token;
  readonly superClass: VariableExpression | null;
  readonly methods: Array<VariableStatement | FunctionStatement>;
  constructor(
    name: Token,
    superClass: VariableExpression | null,
    methods: Array<VariableStatement | FunctionStatement>,
  ) {
    this.name = name;
    this.superClass = superClass;
    this.methods = methods;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitClassStatement(this);
  }
  toString() {
    return `class ${this.name.toString()}{${this.methods
      .map((item) => {
        const temp = item.toString();
        const index = temp.indexOf(' ');
        return temp.slice(index + 1);
      })
      .join('')}}`;
  }
}
export class ExpressionStatement implements Statement {
  readonly expression: Expression;
  constructor(expression: Expression) {
    this.expression = expression;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitExpressionStatement(this);
  }
  toString() {
    return this.expression.toString() + ';';
  }
}
export class FunctionStatement implements Statement {
  readonly name: Token;
  readonly body: BlockStatement;
  readonly params: Token[];
  readonly static: boolean;
  constructor(
    name: Token,
    body: BlockStatement,
    params: Token[],
    isStatic = false,
  ) {
    this.name = name;
    this.body = body;
    this.params = params;
    this.static = isStatic;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitFunctionStatement(this);
  }
  toString() {
    return `function ${
      this.static ? 'static ' : ''
    }${this.name.toString()}(${this.params
      .map((item) => item.toString())
      .join(',')})${this.body.toString()}`;
  }
}
export class IfStatement implements Statement {
  readonly condition: Expression;
  readonly thenBranch: Statement;
  readonly elseBranch: Statement | null;
  constructor(
    condition: Expression,
    thenBranch: Statement,
    elseBranch: Statement | null,
  ) {
    this.condition = condition;
    this.thenBranch = thenBranch;
    this.elseBranch = elseBranch;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitIfStatement(this);
  }
  toString() {
    const temp = `if(${this.condition.toString()})${this.thenBranch.toString()}`;
    if (this.elseBranch === null) {
      return temp;
    }
    return `${temp} else ${this.elseBranch.toString()}`;
  }
}
export class ReturnStatement implements Statement {
  readonly keyword: Token;
  readonly value: Expression | null;
  constructor(keyword: Token, value: Expression | null) {
    this.keyword = keyword;
    this.value = value;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitReturnStatement(this);
  }
  toString() {
    return `return ${this.value === null ? '' : this.value.toString()};`;
  }
}
export class VariableStatement implements Statement {
  readonly name: Token;
  readonly initializer: Expression | null;
  readonly static: boolean;
  constructor(name: Token, initializer: Expression | null, isStatic = false) {
    this.name = name;
    this.initializer = initializer;
    this.static = isStatic;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitVariableStatement(this);
  }
  toString() {
    const temp = `var ${this.static ? 'static ' : ''}${this.name.toString()}`;
    if (this.initializer === null) {
      return temp + ';';
    }
    return `${temp} = ${this.initializer.toString()};`;
  }
}
export class WhileStatement implements Statement {
  readonly condition: Expression;
  readonly body: Statement;
  constructor(condition: Expression, body: Statement) {
    this.condition = condition;
    this.body = body;
  }
  accept(visitor: StatementVisitor): LiteralType {
    return visitor.visitWhileStatement(this);
  }
  toString() {
    return `while(${this.condition.toString()})${this.body.toString()}`;
  }
}

File: statement.go
package main

type BlockStatement struct {
	statements []Statement
}

func (blockStatement BlockStatement) accept(visitor StatementVisitor) any {
	return visitor.visitBlockStatement(blockStatement)
}

type ClassStatement struct {
	name       *Token
	superClass VariableExpression
	methods    []FunctionStatement
}

func (classStatement ClassStatement) accept(visitor StatementVisitor) any {
	return visitor.visitClassStatement(classStatement)
}

type ExpressionStatement struct {
	expression Expression
}

func (expressionStatement ExpressionStatement) accept(visitor StatementVisitor) any {
	return visitor.visitExpressionStatement(expressionStatement)
}

type FunctionStatement struct {
	name   *Token
	body   BlockStatement
	params []*Token
}

func (functionStatement FunctionStatement) accept(visitor StatementVisitor) any {
	return visitor.visitFunctionStatement(functionStatement)
}

type IfStatement struct {
	condition  Expression
	thenBranch Statement
	elseBranch Statement
}

func (ifStatement IfStatement) accept(visitor StatementVisitor) any {
	return visitor.visitIfStatement(ifStatement)
}

type PrintStatement struct {
	expression Expression
	comment    *Token
}

func (printStatement PrintStatement) accept(visitor StatementVisitor) any {
	return visitor.visitPrintStatement(printStatement)
}

type ReturnStatement struct {
	keyword *Token
	value   Expression
}

func (returnStatement ReturnStatement) accept(visitor StatementVisitor) any {
	return visitor.visitReturnStatement(returnStatement)
}

type VariableStatement struct {
	name        *Token
	initializer Expression
}

func (variableStatement VariableStatement) accept(visitor StatementVisitor) any {
	return visitor.visitVariableStatement(variableStatement)
}

type WhileStatement struct {
	condition Expression
	body      Statement
}

func (whileStatement WhileStatement) accept(visitor StatementVisitor) any {
	return visitor.visitWhileStatement(whileStatement)
}

type StatementVisitor interface {
	visitBlockStatement(statement BlockStatement) any
	visitClassStatement(statement ClassStatement) any
	visitExpressionStatement(statement ExpressionStatement) any
	visitFunctionStatement(statement FunctionStatement) any
	visitIfStatement(statement IfStatement) any
	visitPrintStatement(statement PrintStatement) any
	visitReturnStatement(statement ReturnStatement) any
	visitVariableStatement(statement VariableStatement) any
	visitWhileStatement(statement WhileStatement) any
}

type Statement interface {
	accept(visitor StatementVisitor) any
}
